### PR TITLE
Block building core

### PR DESCRIPTION
**Pull request by danielSanchezQ**
* Added transactions selections trait
* Added blob selection trait
* Implemented basic size filling tx selector
* Implemented basic size filling blob selector
* Implemented block builder with composing tx+blob selectors

### CODE REVIEW

File: nomos-core/src/block/builder.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
+// std
+use serde::de::DeserializeOwned;
+use serde::Serialize;
```
Comment (Zeegomo) [inline nomos-core/src/block/builder.rs:3]: nit: not std

File: nomos-core/src/block/builder.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+// crates
+
+// internal
+use crate::block::Block;
+use crate::da::blob::{Blob, BlobSelect};
+use crate::tx::{Transaction, TxSelect};
+use consensus_engine::overlay::RandomBeaconState;
+use consensus_engine::{NodeId, Qc, View};
+
+pub struct BlockBuilder<Tx, Blob, TxSelector, BlobSelector> {
```
Comment (Zeegomo) [inline nomos-core/src/block/builder.rs:15]: how would this be used? I think an example might help me understand

File: nomos-core/src/block/builder.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    _tx: PhantomData<Tx>,
+    _blob: PhantomData<Blob>,
```
Comment (al8n) [inline nomos-core/src/block/builder.rs:17]: It seems not need `PhantomData<_>` here as we have `Item = ...`.

Comment (danielSanchezQ) [reply nomos-core/src/block/builder.rs:17]: Oh yes, this is a leftover from early writing, thanks!

File: nomos-core/src/block/builder.rs (hunk 2 above)
Comment (danielSanchezQ) [reply nomos-core/src/block/builder.rs:31]: I'm not really sure I understand exactly. But it is just a wrapper over the `Block::new` that throws into the mix the transaction and blob selectors (Which should be some static heuristic). 

Core is the build method:

```rust
Block::new(
                view,
                parent_qc,
                tx_selector.select_tx_from(txs),
                blob_selector.select_blob_from(blobs),
                proposer,
                beacon,
            )
```

The idea is to just use it when the leader propose the block.

I had in mind that we could even have a service that handles the next block building depending on different events in the node:
* View change
* New tx arrived
* New blob arrived
* ...
So it always have something ready when needed. 

But for now I think this is enough.

Comment (danielSanchezQ) [reply nomos-core/src/block/builder.rs:31]: > how would this be used? I think an example might help me understand

I added an example to the builder, but I do not think that is what you reffered to, hence my answer above.

File: nomos-core/src/utils/select.rs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+pub fn select_from_till_fill_size<'i, const SIZE: usize, T>(
+    mut measure: impl FnMut(&T) -> usize + 'i,
+    items: impl Iterator<Item = T> + 'i,
+) -> Box<dyn Iterator<Item = T> + 'i> {
+    let mut current_size = 0usize;
+    Box::new(items.take_while(move |tx: &T| {
+        current_size += measure(tx);
+        current_size <= SIZE
+    }))
```
Comment (youngjoon-lee) [inline nomos-core/src/utils/select.rs:9]:
```suggestion
    Box::new(items.take_while(move |item: &T| {
        current_size += measure(item);
        current_size <= SIZE
    }))
```
I think the `item` can be anything, not just tx.

File: nomos-core/src/da/blob/select.rs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    fn select_blob_from<'i, I: Iterator<Item = Self::Blob> + 'i>(
+        &self,
+        blobs: I,
+    ) -> Box<dyn Iterator<Item = Self::Blob> + 'i> {
+        utils::select::select_from_till_fill_size::<SIZE, Self::Blob>(
+            |tx| tx.as_bytes().len(),
+            blobs,
+        )
+    }
+}
```
Comment (youngjoon-lee) [inline nomos-core/src/da/blob/select.rs:34]:
```suggestion
#[derive(Default)]
pub struct FillSize<const SIZE: usize, B> {
    _blob: PhantomData<B>,
}

impl<const SIZE: usize, B> FillSize<SIZE, B> {
    pub fn new() -> Self {
        Self {
            _blob: Default::default(),
        }
    }
}

impl<const SIZE: usize, B: Blob> BlobSelect for FillSize<SIZE, B> {
    type Blob = B;

    fn select_blob_from<'i, I: Iterator<Item = Self::Blob> + 'i>(
        &self,
        blobs: I,
    ) -> Box<dyn Iterator<Item = Self::Blob> + 'i> {
        utils::select::select_from_till_fill_size::<SIZE, Self::Blob>(
            |blob| blob.as_bytes().len(),
            blobs,
        )
    }
}
```
I guess these are typos.

**Review by youngjoon-lee — APPROVED** [review]
👍

**Review by youngjoon-lee — APPROVED** [review]

**Pull request merged**